src/main/java/org/hashids/Hashids.java

```diff
@@ -23,154 +29,164 @@
-        seps += alphabet.substring(0, diff);
-        alphabet = alphabet.substring(diff);
+      if (seps_len > seps.length) {
+        final int diff = seps_len - seps.length;
+        seps = concatenate(seps, alphabet, 0, diff);
+        alphabet = copyOfRange(alphabet, diff, alphabet.length);
       } else {
-        seps = seps.substring(0, seps_len);
+        seps = copyOf(seps, seps_len);
       }
     }
 
-    alphabet = Hashids.consistentShuffle(alphabet, this.salt);
+    alphabet = Hashids.consistentShuffle(alphabet, salt);
     // use double to round up
-    final int guardCount = (int) Math.ceil((double) alphabet.length() / GUARD_DIV);
+    final int guardCount = (int) Math.ceil((double) alphabet.length / GUARD_DIV);
 
-    String guards;
-    if (alphabet.length() < 3) {
-      guards = seps.substring(0, guardCount);
-      seps = seps.substring(guardCount);
+    char[] guards;
+    if (alphabet.length < 3) {
+      guards = copyOf(seps, guardCount);
+      seps = copyOfRange(seps, guardCount, seps.length);
     } else {
-      guards = alphabet.substring(0, guardCount);
-      alphabet = alphabet.substring(guardCount);
+      guards = copyOf(alphabet, guardCount);
+      alphabet = copyOfRange(alphabet, guardCount, alphabet.length);
     }
+
     this.guards = guards;
     this.alphabet = alphabet;
     this.seps = seps;
+    this.validChars = concatenate(alphabet, guards, seps);
+    this.guardsRegExp = '[' + String.valueOf(guards) + ']';
+    this.sepsRegExp = '[' + String.valueOf(seps) + ']';
+  }
+
+  private void validateAlphabet(char[] alphabet) {
+    if (alphabet.length < MIN_ALPHABET_LENGTH) {
+      throw new IllegalArgumentException(
+          "The alphabet must contain at least " + MIN_ALPHABET_LENGTH + " unique characters.");
+    }
+
+    for (int i = 0; i < alphabet.length; i++) {
+      if (isSpaceChar(alphabet[i])) {
+        throw new IllegalArgumentException("The alphabet cannot contain spaces.");
+      }
+
+      if ((i + 1 < alphabet.length) && (alphabet[i] == alphabet[i + 1])) {
+        throw new IllegalArgumentException("The alphabet cannot contain duplicates.");
+      }
+    }
   }
 
+  // ---------
+
   /**
    * Encode numbers to string
    *
-   * @param numbers
-   *          the numbers to encode
+   * @param numbers the numbers to encode
    * @return the encoded string
    */
   public String encode(long... numbers) {
     if (numbers.length == 0) {
-      return "";
+      throw new IllegalArgumentException("At least one number must be specified.");
     }
 
     for (final long number : numbers) {
       if (number < 0) {
-        return "";
+        return ""; // we must throw an exception here (like the case when we compare with MAX_NUMBER)
       }
+
       if (number > MAX_NUMBER) {
-        throw new IllegalArgumentException("number can not be greater than " + MAX_NUMBER + "L");
+        throw new IllegalArgumentException("Number can not be greater than " + MAX_NUMBER + '.');
```

Comment:
I know, but the API it already shows to the user that the parameter is a long. Also, Java doesn't  allow us to use another type, that is not compatible with this one (e.g. `float`).

Reply:
But if you try to assign a long literal without L it will not work.

Reply:
Then it is not a `long`. The parameter type is very clear: `long... numbers`. 

Placing an `L` on the error it doesn't help because, before anything else, the compiler will do its job.